---
title: "Ribo-seq preprocessing (by riboWaltz)"
author: "Chao Di, dic@email.chop.edu"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc_depth: 1
    number_sections: true
---

```{r setup, include=FALSE, echo = FALSE, cache = FALSE}
# replace with path where you want the results be
knitr::opts_knit$set(root.dir="/Users/dic/Documents/Will Bailis lab/GSE120762_riboseq")
```

# Abstract
This script is made from `riboWaltz` package for ribo-seq preprocessing, including P site identification and related stats. Codon usage is also analyzed. Most of the instructions are directly adapted from the [Github](https://github.com/LabTranslationalArchitectomics/riboWaltz), only for personal usage.
riboWaltz currently works for read alignments based on transcript coordinates. This choice is due to the main purpose of RiboSeq assays to study translational events through the isolation and sequencing of ribosome protected fragments. Most reads from RiboSeq are supposed to map on mRNAs and not on introns and intergenic regions. BAM based on transcript coordinates can be generated in two ways: i) aligning directly against transcript sequences; ii) aligning against standard chromosome sequences, requiring the outputs to be translated in transcript coordinates.

```{r include=FALSE,echo=FALSE,message = FALSE, warning=FALSE}
library(riboWaltz)
library(ggplot2)
library(ggpubr)
```

## Annoation data table
A reference annotation file is required by many functions included in the package. The annotation file must be a data table of at least five columns reporting, for each reference transcript, its name, its length and the length of its annotated 5' UTR, CDS and 3'UTR. Please note: if a transcript region is not annotated its length is set to 0. The annotation file can be either provided by the user or generated starting from GTF files or TxDb objects by running the function create_annotation. Please make sure the GTF or the TxDb object the derives from the same release of the sequences used in the alignment step. Here is an example of the output:


```{r include=TRUE,echo=FALSE,message = FALSE, warning=FALSE}
data.table::data.table(
  transcript = c("ENSMUST00000000001.4",
                 "ENSMUST00000000003.11","ENSMUST00000000010.8"),
        l_tr = c(3262L, 902L, 2574L),
      l_utr5 = c(141L, 140L, 85L),
       l_cds = c(1065L, 525L, 753L),
      l_utr3 = c(2056L, 237L, 1736L)
)
# setwd("/Users/dic/Documents/Will Bailis lab/GSE120762_riboseq")
annotation_dt <- create_annotation(gtfpath="./mm10.refGene.gtf", dataSource="UCSC", organism="Mus musculus")
```

## convert BAM files to data tables
One or multiple BAM files are read, converted into data tables or GRanges objects and arranged in a list or a GRangesList, respectively, by running function `bamtolist`.
The resulting data tables contain, for each read: i) the name of the corresponding reference sequence (i.e. of the transcript on which it aligns); ii) its leftmost and rightmost position with respect to the 1st nucleotide of the reference sequence; iii) its length; iv) the leftmost and rightmost position of the annotated CDS of the reference sequence (if any) with respect to its 1st nucleotide. 

```{r include=TRUE, echo=FALSE, message = FALSE, warning=FALSE}
# example table
data.table::data.table(
  transcript = c("ENSMUST00000000001.4",
                 "ENSMUST00000000001.4","ENSMUST00000000001.4","ENSMUST00000000001.4"),
        end5 = c(92L, 94L, 131L, 132L),
        end3 = c(119L, 122L, 159L, 158L),
      length = c(28L, 29L, 29L, 27L),
   cds_start = c(142L, 142L, 142L, 142L),
    cds_stop = c(1206L, 1206L, 1206L, 1206L)
)
# load data
reads_list <- bamtolist(bamfolder="STAR_align_transcriptome/", annotation=annotation_dt)
# filter by read length
# filtered_list <- length_filter(data = reads_list, length_filter_mode = "custom", length_filter_vector = 26:32)
```

## P-site offset
The P-site offsets (POs) are computed by the function psite starting from a list of data tables generated by bamtolist, bedtolist or length_filter. The PO is defined as the distance between the extremities of a read and the first nucleotide of the P-site itself (see more details in the [paper](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1006169)). 
`psite` returns a data table containing, for all samples and read lengths: i) the percentage of reads in the whole dataset, ii) the percentage of reads aligning on the start codon (if any); iii) the distance of the P-site from the two extremities of the reads before and after the correction step; iv) the name of the sample:

```{r include=TRUE,echo=FALSE,message = FALSE, warning=FALSE}
# example 
# data.table::data.table(
#                    length = c(19L, 20L, 21L, 22L),
#          total_percentage = c(0.715, 0.811, 0.992, 0.921),
#          start_percentage = c(0.11, 0.194, 0.223, 0.364),
#              around_start = c(TRUE, TRUE, TRUE, TRUE),
#             offset_from_5 = c(12L, 12L, 15L, 6L),
#             offset_from_3 = c(6L, 7L, 5L, 15L),
#   corrected_offset_from_5 = c(8L, 7L, 7L, 10L),
#   corrected_offset_from_3 = c(10L, 12L, 13L, 11L),
#                    sample = c("Samp1", "Samp1", "Samp1", "Samp1")
# )
# P-site offset calculation
POs <- psite(reads_list, flanking=6, extremity="auto")
write.table(POs, file="P-site_offset.tsv",quote=F, sep="\t", row.names=FALSE)
POs
```
## Update reads table with Psite info
`psite_info` attaches to each data table four columns reporting i) the P-site position with respect to the 1st nucleotide of the transcript, ii) the P-site position with respect to the start and the stop codon of the annotated coding sequence (if any) and iii) the region of the transcript (5' UTR, CDS, 3' UTR) that includes the P-site. Please note: for transcripts not associated to any annotated CDS the position of the P-site with respect to the start and the stop codon is set to NA. Here an example:

```{r include=TRUE,echo=FALSE,message = FALSE, warning=FALSE}
# example
data.table::data.table(
        transcript = c("ENSMUST00000000001.4",
                       "ENSMUST00000000001.4","ENSMUST00000000001.4",
                       "ENSMUST00000000001.4","ENSMUST00000000001.4"),
              end5 = c(92L, 94L, 131L, 132L, 140L),
             psite = c(103L, 106L, 143L, 142L, 151L),
              end3 = c(119L, 122L, 159L, 158L, 167L),
            length = c(28L, 29L, 29L, 27L, 28L),
         cds_start = c(142L, 142L, 142L, 142L, 142L),
          cds_stop = c(1206L, 1206L, 1206L, 1206L, 1206L),
  psite_from_start = c(-39L, -36L, 1L, 0L, 9L),
   psite_from_stop = c(-1103L, -1100L, -1063L, -1064L, -1055L),
      psite_region = c("5utr", "5utr", "cds", "cds", "cds")
)
# update reads_list with p-site info
reads_psite_list <- psite_info(reads_list, POs)
```

# Plots to view the data
RPF (ribosome protected fragment) length distribution, P-sites distribution on three reading frames for different regions (5'UTR, CDS, 3'UTR), metagene for trinucleotide periodicity on CDSs

## RPF length distribution
The first plot, provided by `rlength_distr`, shows the distribution of reads lengths for a specified sample. It can be exploited for i) identifying multiple populations of read length, associated to different ribosome conformations and ii) exploring the contribution of each length bin to the final P-site determination.

```{r include=TRUE,fig.height=12, fig.width=14,echo=FALSE,message = FALSE, warning=FALSE}

all_samples = names(reads_list)
figs=list()
for(sample_name in all_samples){
  length_dist <- rlength_distr(reads_list, sample=sample_name)
  figs <- c(figs, list(length_dist[["plot"]]))
}
ggarrange(plotlist=figs, nrow=2, ncol=2)
```

## metaheatmaps displays the abundance of the 5' and 3' end of reads on CDSs
The second plot, provided by rends_heat, consists of four metaheatmaps. It displays the abundance of the 5' and 3' extremity of reads mapping on and around the start and the stop codon of annotated CDSs, stratified by their length. 

```{r include=TRUE,fig.height=10, fig.width=20,echo=FALSE,message = FALSE, warning=FALSE}
figs=list()
for(sample_name in all_samples){
  ends_heatmap <- rends_heat(reads_list, annotation_dt, sample=sample_name, utr5l = 25, cdsl = 50, utr3l = 25)
  figs <- c(figs, list(ends_heatmap[["plot"]] + theme(plot.title=element_text(size=20))))
}
ggarrange(plotlist=figs, nrow=2, ncol=2)

```

## P-sites per region (5’UTRs, CDSs and 3’UTRs)
Ribosome profiling data should highlight the CDS of transcripts as the region with the higher percentage of reads. To verify this property the function region_psite computes the percentage of P-sites falling in the three annotated transcript regions (5' UTR, CDS and 3' UTR). The bar plot of the resulting values includes a bar called "RNAs" displaying the expected read distribution from a random fragmentation of RNA. It reports the percentage of regions length for the transcripts included in the analysis, based on the cumulative nucleotide length of 5' UTRs, CDSs and 3' UTRs. 

```{r include=TRUE,fig.height=12, fig.width=14,echo=FALSE,message = FALSE, warning=FALSE}
figs=list()
for(sample_name in all_samples){
  psite_region <- region_psite(reads_psite_list, annotation_dt, sample=sample_name)
  figs <- c(figs, list(psite_region[["plot"]]))
}
ggarrange(plotlist=figs, nrow=2, ncol=2)
```

## P-sites/RPF signal on three reading frames for 5’UTRs, CDSs and 3’UTRs
Compute the percentage of P-sites falling in the three possible translation reading frames for 5’ UTRs, CDSs and 3’ UTRs, either analyses all read lengths separately and generates a heatmap for each transcript region, or processes all reads at once, returning three bar plots.
### by read length
```{r include=TRUE,fig.height=12, fig.width=14,echo=FALSE,message = FALSE, warning=FALSE}
figs=list()
for(sample_name in all_samples){
  frames_stratified <- frame_psite_length(reads_psite_list, sample=sample_name, region = "all")
  figs <- c(figs, list(frames_stratified[["plot"]]))
}
ggarrange(plotlist=figs, nrow=2, ncol=2)
```

### in total

```{r include=TRUE,fig.height=12, fig.width=14,echo=FALSE,message = FALSE, warning=FALSE}
figs=list()
for(sample_name in all_samples){
  frames <- frame_psite(reads_psite_list, sample =sample_name, region = "all")
  figs <- c(figs, list(frames[["plot"]]))
}
ggarrange(plotlist=figs, nrow=2, ncol=2)
```

## metaplot to show trinucleotide periodicity along CDSs
Metaplot to show the trinucleotide periodicity along coding sequences. It generates metaprofiles based on P-sites mapping around the start and the stop codon of annotated CDSs. 

```{r include=TRUE,fig.height=12, fig.width=22,echo=FALSE,message = FALSE, warning=FALSE}
figs=list()
for(sample_name in all_samples){
  metaprofile <- metaprofile_psite(reads_psite_list, annotation_dt, sample=sample_name, utr5l=20, cdsl=50, utr3l=20,    plot_title="sample.transcript")
  figs <- c(figs, list(metaprofile[[paste0("plot_",sample_name)]] + theme(plot.title=element_text(size=20))))
}
ggarrange(plotlist=figs, nrow=2, ncol=2)
```

## Codon usage
To understand what codons display higher or lower ribosome density, the codon usage index is defined as the frequency of in-frame P-sites along the coding sequence codon by codon optionally normalized for the frequency in sequences of each codon. A bar plot is highlighting the start and stop codons and labelling each bar with the corresponding amino acid.
```{r include=TRUE,fig.height=12, fig.width=22,echo=FALSE,message = FALSE, warning=FALSE}
figs=list()
for(sample_name in all_samples){
  cu_barplot <- codon_usage_psite(reads_psite_list, annotation_dt, sample = sample_name,
                                        fastapath = "./refGene_transcript.fa",
                                        fasta_genome = FALSE,
                                        frequency_normalization = TRUE) 
  figs <- c(figs, list(cu_barplot[["plot"]] + ggtitle(sample_name) +  theme(plot.title=element_text(size=20))))
}
ggarrange(plotlist=figs, nrow=2, ncol=2)
```
